### ISSUE TITLE
io.Pipe{Reader,Writer}.Close resets error code set with CloseWithError

### ISSUE THREAD

**Issue by cyphar** [golang/go#17992]
```sh
% go version
go version go1.7 linux/amd64
% go env
GOARCH="amd64"
GOBIN="/home/cyphar/go/bin"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/cyphar/.local"
GORACE=""
GOROOT="/usr/lib64/go"
GOTOOLDIR="/usr/lib64/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build573484999=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
```

If you use `io.Pipe` you'll notice that calling `.Close` on the writer will un-do the error code set by `.CloseWithError(err)`. I'm unsure if this is a bug, but I found this to be very unexpected. In particular when it's usual for you to use a `defer writer.Close()` code style. This is also available on play.golang.org: https://play.golang.org/p/zJDmmRTOLQ.

```go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
)

func main() {
	reader, writer := io.Pipe()

	go func() {
		defer writer.Close()

		fmt.Fprintf(writer, "some data to write before the error")
		writer.CloseWithError(fmt.Errorf("an error code"))
		fmt.Fprintf(writer, "some write that will fail")
	}()

	val, err := ioutil.ReadAll(reader)
	fmt.Printf("val: %s\n", val)
	fmt.Printf("err: %s\n", err)
}
```

If you comment out `defer writer.Close()` you see what I would expect.

### What did you expect to see?
```
% go run pipe.go
val: some data to write before the error
err: an error code
```

### What did you see instead?
```
% go run pipe.go
val: some data to write before the error
err: %!s(<nil>)
```

In particular, I did not expect for the error returned to change to `EOF`.

**Comment by cyphar** [comment]
My proposed patch looks like this:

```diff
commit f3bce6cef511a1c64fb668fd452b2a5c364e3dc9
Author: Aleksa Sarai <cyphar@cyphar.com>
Date:   Mon Nov 21 02:23:53 2016 +1100

    io: pipe: make CloseWithError final

    CloseWithError would previously overwrite previously set error flags,
    making a defer'd pipe.Close() set the error to EOF (clearing the
    previous error flags). This was quite confusing to users, because it
    meant that io.Pipes we're very different to channels (double-closing a
    pipe would not produce an error and would actually affect the pipe).

    Fixes: golang/go#17992
    Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

diff --git a/src/io/pipe.go b/src/io/pipe.go
index 7e98cd2eb788..a462e857493c 100644
--- a/src/io/pipe.go
+++ b/src/io/pipe.go
@@ -93,23 +93,29 @@ func (p *pipe) write(b []byte) (n int, err error) {
        return
 }

-func (p *pipe) rclose(err error) {
+func (p *pipe) rclose(err error) error {
        if err == nil {
                err = ErrClosedPipe
        }
        p.l.Lock()
        defer p.l.Unlock()
+       if p.rerr != nil {
+               return ErrClosedPipe
+       }
        p.rerr = err
        p.rwait.Signal()
        p.wwait.Signal()
 }

-func (p *pipe) wclose(err error) {
+func (p *pipe) wclose(err error) error {
        if err == nil {
                err = EOF
        }
        p.l.Lock()
        defer p.l.Unlock()
+       if p.werr != nil {
+               return ErrClosedPipe
+       }
        p.werr = err
        p.rwait.Signal()
        p.wwait.Signal()
@@ -138,8 +144,7 @@ func (r *PipeReader) Close() error {
 // CloseWithError closes the reader; subsequent writes
 // to the write half of the pipe will return the error err.
 func (r *PipeReader) CloseWithError(err error) error {
-       r.p.rclose(err)
-       return nil
+       return r.p.rclose(err)
 }

 // A PipeWriter is the write half of a pipe.
@@ -168,8 +173,7 @@ func (w *PipeWriter) Close() error {
 //
 // CloseWithError always returns nil.
 func (w *PipeWriter) CloseWithError(err error) error {
-       w.p.wclose(err)
-       return nil
+       return w.p.wclose(err)
 }

 // Pipe creates a synchronous in-memory pipe.
```

**Comment by bradfitz** [comment]
Double-Close isn't really defined either way, but reading the existing docs in the io package, it looks like the existing behavior matches what the docs say, even if it wasn't what you expected.

Changing it seems unwise. It seems just as likely to surprise people in the other direction and break existing code in the wild, so I'm inclined to not change behavior.
